File: frontend/src/lib/journal.ts
import { api } from './api'

export type Mood = '' | 'good' | 'flat' | 'low' | 'energised' | 'drained'
export const MOODS: Exclude<Mood, ''>[] = ['good', 'flat', 'low', 'energised', 'drained']

export interface JournalEntry {
  id: string
  title: string
  body: string
  mood: Mood
  tags: string[]
  linked_objective_id: string | null
  linked_project_id: string | null
  learn_from_style: boolean
  created_at: string
  updated_at: string
}

export type JournalDraft = Partial<
  Pick<
    JournalEntry,
    | 'title'
    | 'body'
    | 'mood'
    | 'tags'
    | 'linked_objective_id'
    | 'linked_project_id'
    | 'learn_from_style'
  >
>

export async function fetchEntries(): Promise<JournalEntry[]> {
  const { data } = await api.get<JournalEntry[]>('/journal')
  return data
}

export async function createEntry(draft: JournalDraft): Promise<JournalEntry> {
  const { data } = await api.post<JournalEntry>('/journal', draft)
  return data
}

export async function patchEntry(id: string, draft: JournalDraft): Promise<JournalEntry> {
  const { data } = await api.patch<JournalEntry>(`/journal/${id}`, draft)
  return data
}

export async function deleteEntry(id: string): Promise<void> {
  await api.delete(`/journal/${id}`)
}

export async function continueEntry(id: string): Promise<string> {
  const { data } = await api.post<{ text: string }>(`/journal/${id}/continue`)
  return data.text
}

export function entryTitle(e: Pick<JournalEntry, 'title' | 'body'>): string {
  return e.title.trim() || e.body.trim().split('\n')[0].slice(0, 60) || 'Untitled'
}

File: backend/app/models/journal.py
"""Journal schemas (Phase 2.5, REQ-1/2/12).

`body` is the substantive field; `title` may be empty. A long entry with
`learn_from_style` set is mirrored into `style_samples` by `journal_store` (J1).
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

JournalMood = Literal["", "good", "flat", "low", "energised", "drained"]

TITLE_MAX = 200
BODY_MAX = 20_000
TAGS_MAX = 12


class JournalCreate(BaseModel):
    title: str = Field(default="", max_length=TITLE_MAX)
    body: str = Field(default="", max_length=BODY_MAX)
    mood: JournalMood = ""
    tags: list[str] = Field(default_factory=list, max_length=TAGS_MAX)
    linked_objective_id: str | None = None
    linked_project_id: str | None = None
    learn_from_style: bool = True


class JournalPatch(BaseModel):
    title: str | None = Field(default=None, max_length=TITLE_MAX)
    body: str | None = Field(default=None, max_length=BODY_MAX)
    mood: JournalMood | None = None
    tags: list[str] | None = Field(default=None, max_length=TAGS_MAX)
    linked_objective_id: str | None = None
    linked_project_id: str | None = None
    learn_from_style: bool | None = None


class JournalOut(BaseModel):
    id: str
    title: str
    body: str
    mood: JournalMood
    tags: list[str]
    linked_objective_id: str | None
    linked_project_id: str | None
    learn_from_style: bool
    created_at: str
    updated_at: str


class ContinueOut(BaseModel):
    text: str
